File: Classes/Graphics and Tools/Text/DrawText.m
/*
DrawText.m
Draw

Copyright © 2021, AJ Raftis and AJRFoundation authors
All rights reserved.

Redistribution and use in source and binary forms, with or without modification,
are permitted provided that the following conditions are met:

* Redistributions of source code must retain the above copyright notice, this 
  list of conditions and the following disclaimer.
* Redistributions in binary form must reproduce the above copyright notice, 
  this list of conditions and the following disclaimer in the documentation 
  and/or other materials provided with the distribution.
* Neither the name of Draw nor the names of its contributors may be 
  used to endorse or promote products derived from this software without 
  specific prior written permission.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED 
WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE 
DISCLAIMED. IN NO EVENT SHALL AJ RAFTIS BE LIABLE FOR ANY DIRECT, INDIRECT, 
INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR 
PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF 
LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE
OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF 
ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
*/

#import "DrawText.h"

#import "DrawEvent.h"
#import "DrawFunctions.h"
#import "DrawPage.h"
#import "DrawTextContainer.h"
#import "DrawTextView.h"
#import "DrawDocument.h"

#import <AJRFoundation/AJRFoundation.h>
#import <AJRInterfaceFoundation/AJRInterfaceFoundation.h>

NSString * const DrawTextIdentifier = @"text";

#if !defined (UseDrawTextContainer)
#define DrawTextContainer NSTextContainer
@interface NSTextContainer (Extensions)
- (void)setGraphic:(DrawGraphic *)graphic;
- (void)graphicDidChangeShape:(DrawGraphic *)graphic;
@end
@implementation NSTextContainer (Extensions)
- (void)setGraphic:(DrawGraphic *)graphic { }
- (void)graphicDidChangeShape:(DrawGraphic *)graphic { }
@end
#endif

@implementation DrawText {
    NSTextStorage *_textStorage;

    BOOL _editing;
    BOOL _ignoreGraphicShapeChange;
}

#pragma mark - Creation

- (void)initializeFromString:(NSAttributedString *)string {
    NSLayoutManager *manager;
    DrawTextContainer *container;
    DrawTextView *textView;
    NSSize size;
    
    _textStorage = [[NSTextStorage alloc] init];
    
    if (string) {
        [self setAttributedString:string];
    }
    
    manager = [[NSLayoutManager alloc] init];
    [_textStorage addLayoutManager:manager];
    
    if (self.graphic) {
        size = [self.graphic frame].size;
        if (size.width < 15.0) size.width = 15.0;
        if (size.height < 15.0) size.height = 15.0;
    } else {
        size = (NSSize){15.0, 15.0};
    }
    _lineFragmentPadding = 2.0;
    container = [[DrawTextContainer alloc] initWithContainerSize:size];
    [container setLineFragmentPadding:_lineFragmentPadding];
    [manager addTextContainer:container];
    [container setGraphic:self.graphic];
    [container setHeightTracksTextView:YES];
    [container setWidthTracksTextView:YES];
    
    textView = [[DrawTextView alloc] initWithFrame:(NSRect){{0.0, 0.0}, size} textContainer:container];
    [textView setDrawsBackground:NO];
    [textView setEditable:NO];
    [textView setSelectable:NO];
    [textView setHorizontallyResizable:NO];
    [textView setVerticallyResizable:NO];
    [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(textViewFrameDidChange:) name:NSViewFrameDidChangeNotification object:textView];
}

- (id)initWithGraphic:(DrawGraphic *)aGraphic {
    if ((self = [super initWithGraphic:aGraphic])) {
        [self initializeFromString:nil];
    }
    return self;
}

- (id)initWithGraphic:(DrawGraphic *)aGraphic text:(NSAttributedString *)someText {
    if ((self = [super initWithGraphic:aGraphic])) {
        [self initializeFromString:someText];
    }
    return self;
}

- (void)prepareTextInLayoutManager {
    NSLayoutManager *layoutManager = [self layoutManager];
    NSRange range = [layoutManager glyphRangeForTextContainer:[self textContainer]];
    [layoutManager ensureLayoutForGlyphRange:range];
}

- (DrawGraphicCompletionBlock)drawPath:(AJRBezierPath *)path withPriority:(DrawAspectPriority)priority {
    if (!_editing) {
        NSLayoutManager *layoutManager = self.layoutManager;
        NSRect frame = self.graphic.frame;
        NSRange range = [layoutManager glyphRangeForTextContainer:self.textContainer];
        
        [layoutManager drawBackgroundForGlyphRange:range atPoint:frame.origin];
        [layoutManager drawGlyphsForGlyphRange:range atPoint:frame.origin];
    }
	
	return NULL;
}

- (void)setAttributedString:(NSAttributedString *)string {
    [self.graphic.document registerUndoWithTarget:self selector:@selector(setAttributedString:) object:[_textStorage copy]];
    [_textStorage setAttributedString:string];
    [self.graphic setNeedsDisplay];
}

- (NSAttributedString *)attributedString {
    return _textStorage;
}

- (void)setLineFragmentPadding:(CGFloat)aPadding {
    if (aPadding != _lineFragmentPadding) {
        NSInteger x;
        NSArray *textContainers = [[[_textStorage layoutManagers] lastObject] textContainers];
        NSTextContainer	*container;
        
        [[[self.graphic document] prepareWithInvocationTarget:self] setLineFragmentPadding:_lineFragmentPadding];
        _lineFragmentPadding = aPadding;
        
        for (x = 0; x < (const NSInteger)[textContainers count]; x++) {
            container = [textContainers objectAtIndex:x];
            [container setLineFragmentPadding:_lineFragmentPadding];
        }
        [self.graphic setNeedsDisplay];
    }
}

- (NSLayoutManager *)layoutManager {
    return [[_textStorage layoutManagers] lastObject];
}

- (DrawTextView *)textView {
    return (DrawTextView *)[(NSTextContainer *)[[[[_textStorage layoutManagers] lastObject] textContainers] lastObject] textView];
}

- (DrawTextContainer *)textContainer {
    return AJRObjectIfKindOfClass(_textStorage.layoutManagers.lastObject.textContainers.lastObject, DrawTextContainer);
}

- (void)setGraphic:(DrawGraphic *)graphic {
    if (self.graphic != graphic) {
        [super setGraphic:graphic];
        [[self textContainer] setGraphic:self.graphic];
    }
}

- (void)updateMaxSize {
    NSPrintInfo	*printInfo = [[self.graphic document] printInfo];
    NSSize paperSize = [printInfo paperSize];
    NSPoint origin = [self.graphic frame].origin;
    NSTextView *textView = [self textView];
    NSSize maxSize;
    
    if (!printInfo) return;
    
    maxSize.width = paperSize.width - [printInfo rightMargin] - origin.x;
    maxSize.height = paperSize.height - [printInfo bottomMargin] - origin.y;
    AJRLogDebug(@"maxSize = %@\n", NSStringFromSize(maxSize));
    [textView setMaxSize:maxSize];
}

- (void)graphicDidAddToDocument:(DrawDocument *)aView {
    [self updateMaxSize];
}

- (void)graphicWillRemoveFromDocument:(DrawDocument *)aView {
}

- (void)graphicDidChangeShape:(DrawGraphic *)aGraphic {
    if (!_ignoreGraphicShapeChange) {
        DrawTextView *textView = [self textView];
        DrawTextContainer *textContainer = [self textContainer];
        NSRect textFrame, graphicFrame;

        textFrame = [textView frame];
        graphicFrame = [self.graphic frame];
        if (!NSEqualRects(textFrame, graphicFrame)) {
            [textView setFrame:graphicFrame];
            //[textContainer graphicDidChangeShape:aGraphic];
        }
        
        if (![textContainer isSimpleRectangularTextContainer]) {
            [textContainer graphicDidChangeShape:aGraphic];
        }
        if (!NSEqualSizes(textFrame.size, graphicFrame.size)) {
            NSSize size = graphicFrame.size;
            if (size.width < 15.0) size.width = 15.0;
            if (size.height < 15.0) size.height = 15.0;
            [textContainer setContainerSize:size];
        }
        [self updateMaxSize];
    }
}

- (BOOL)aspectAcceptsEdit {
    return YES;
}

- (void)setupTextView:(NSTextView *)textView {
    NSRect frame = [self.graphic frame];
    NSRect bounds = frame;
    NSTextContainer	*container = [self textContainer];
    
    [textView setFrame:frame];
    [textView setBounds:bounds];
    [textView setEditable:YES];
    [textView setSelectable:YES];
    [textView setHorizontallyResizable:YES];
    [textView setVerticallyResizable:YES];
    [container setHeightTracksTextView:YES];
    [container setWidthTracksTextView:YES];
    [self updateMaxSize];
    [textView sizeToFit];
}

- (BOOL)beginEditingFromEvent:(DrawEvent *)event {
    DrawTextView *textView = [self textView];
    NSLayoutManager *layoutManager = [self layoutManager];
    NSPoint point, origin;
    CGFloat distance;
    NSRange range = {0, 0};
    NSRect frame, bounds;
    
    frame = [self.graphic frame];
    bounds = (NSRect){{0.0, 0.0}, frame.size};
    [self setupTextView:textView];
    [[self.graphic page] addSubview:textView];
    [[textView window] makeFirstResponder:textView];
    origin = [self.graphic frame].origin;
    point = event.locationOnPage;
    point.x -= origin.x;
    point.y -= origin.y;
    range.location = [layoutManager glyphIndexForPoint:point inTextContainer:[self textContainer] fractionOfDistanceThroughGlyph:&distance];
    if (distance > 0.5) {
        range.location++;
    }
    [textView setSelectedRange:range];
    
    _editing = YES;
    
    return YES;
}

- (void)endEditing {
    DrawTextView *textView = self.textView;
    
    if (_editing) {
        textView.editable = NO;
        textView.selectable = NO;
        textView.horizontallyResizable = NO;
        textView.verticallyResizable = NO;
        [textView removeFromSuperview];
        
        _editing = NO;
        
        [NSNotificationCenter.defaultCenter postNotificationName:DrawObjectDidResignRulerNotification object:self.graphic.document];
    }
}

- (BOOL)isPoint:(NSPoint)aPoint inPath:(AJRBezierPath *)path {
    return [path isHitByPoint:aPoint];
}

- (void)textViewFrameDidChange:(NSNotification *)notification {
    //NSLog(@"Change\n");
}

#pragma mark - DrawAspect

+ (DrawAspect *)defaultAspectForGraphic:(DrawGraphic *)graphic {
    return [[self alloc] initWithGraphic:graphic];
}

#pragma mark - NSCopying

- (NSTextStorage *)_createTextStorageWithString:(NSAttributedString *)string {
    return [self _finishInitializingTextStorage:[[NSTextStorage alloc] init] with:string];
}

- (NSTextStorage *)_finishInitializingTextStorage:(NSTextStorage *)storage with:(NSAttributedString *)string {
    [storage setAttributedString:string];
    NSLayoutManager *manager = [[NSLayoutManager alloc] init];
    [storage addLayoutManager:manager];
    DrawTextContainer *container = [[DrawTextContainer alloc] initWithContainerSize:(NSSize){100, 100}];
    [container setLineFragmentPadding:_lineFragmentPadding];
    [container setHeightTracksTextView:YES];
    [container setWidthTracksTextView:YES];
    [manager addTextContainer:container];
    
    return storage;
}

- (id)copyWithZone:(NSZone *)zone {
    DrawText *new = [super copyWithZone:zone];
    
    // NOTE: Apparently copying an NSTextStorage produces an attributed string, not a text storage.
    new->_textStorage = [self _createTextStorageWithString:[_textStorage copy]];
    new->_editing = _editing;
    new->_lineFragmentPadding = _lineFragmentPadding;
    
    return new;
}

#pragma mark - AJRXMLArchiving

+ (NSString *)ajr_nameForXMLArchiving {
    return @"text";
}

- (void)decodeWithXMLCoder:(AJRXMLCoder *)coder {
    [super decodeWithXMLCoder:coder];

    [coder decodeObjectForKey:@"text" setter:^(id  _Nonnull object) {
        // Make sure the input is actually an attributed string, since it might not be if someone fiddled with the document.
        NSAttributedString *string = AJRObjectIfKindOfClass(object, NSAttributedString);
        if (string != nil) {
            self->_textStorage = [self _createTextStorageWithString:string];
        } else {
            self->_textStorage = [self _createTextStorageWithString:[[NSAttributedString alloc] init]];
        }
    }];
    [coder decodeBoolForKey:@"editing" setter:^(BOOL value) {
        self->_editing = value;
    }];
    [coder decodeFloatForKey:@"lineFragmentPadding" setter:^(float value) {
        self->_lineFragmentPadding = value;
    }];
}

- (id)finalizeXMLDecodingWithError:(NSError * _Nullable __autoreleasing *)error {
    CGFloat save = _lineFragmentPadding;

    [self initializeFromString:_textStorage];
    [self setLineFragmentPadding:0];
    [self setLineFragmentPadding:save];

    return self;
}

- (void)encodeWithXMLCoder:(AJRXMLCoder *)encoder {
    [super encodeWithXMLCoder:encoder];

    [encoder encodeObject:_textStorage forKey:@"text"];
	[encoder encodeBool:_editing forKey:@"editing"];
    [encoder encodeFloat:_lineFragmentPadding forKey:@"lineFragmentPadding"];
}

@end

File: Classes/Graphics and Tools/Text/DrawText.swift
/*
 DrawText.swift
 Draw

 Copyright © 2022, AJ Raftis and Draw authors
 All rights reserved.

 Redistribution and use in source and binary forms, with or without modification,
 are permitted provided that the following conditions are met:

 * Redistributions of source code must retain the above copyright notice, this
   list of conditions and the following disclaimer.
 * Redistributions in binary form must reproduce the above copyright notice,
   this list of conditions and the following disclaimer in the documentation
   and/or other materials provided with the distribution.
 * Neither the name of Draw nor the names of its contributors may be
   used to endorse or promote products derived from this software without
   specific prior written permission.

 THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
 ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
 WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
 DISCLAIMED. IN NO EVENT SHALL AJ RAFTIS BE LIABLE FOR ANY DIRECT, INDIRECT,
 INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
 LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
 PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
 LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE
 OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF
 ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

import AJRInterface

let DrawTextIdentifier = "text"

public extension AJRUserDefaultsKey {
    static var debugTextContainerFrames : AJRUserDefaultsKey<Bool> {
        return AJRUserDefaultsKey<Bool>.key(named: "debugTextContainerFrames", defaultValue: false)
    }
}

@objcMembers
open class DrawText : DrawAspect {

    @objc(DrawVerticalTextAlignment)
    public enum VerticalTextAlignment : Int, AJRXMLEncodableEnum {

        case top = 0
        case middle
        case bottom

        public var description: String {
            switch self {
            case .top: return "top"
            case .middle: return "middle"
            case .bottom: return "bottom"
            }
        }
    }

    open var textStorage = NSTextStorage()
    open var size = NSSize.zero
    open var lineFragmentPadding : CGFloat = 2.0 {
        willSet {
            let current = lineFragmentPadding
            graphic?.document?.registerUndo(withTarget: self, handler: { value in
                self.lineFragmentPadding = current
            })
        }
        didSet {
            if let textContainers = textStorage.layoutManagers.last?.textContainers {
                for container in textContainers {
                    container.lineFragmentPadding = lineFragmentPadding
                }
                graphic?.setNeedsDisplay()
            }
        }
    }
    open var verticalAlignment : VerticalTextAlignment = .top {
        didSet {
            graphic?.setNeedsDisplay()
        }
    }

    open var layoutManager : NSLayoutManager? {
        return textStorage.layoutManagers.last
    }

    open var textView : DrawTextView? {
        return layoutManager?.textContainers.last?.textView as? DrawTextView
    }

    open var textContainer : DrawTextContainer? {
        return layoutManager?.textContainers.last as? DrawTextContainer
    }

    open var editing : Bool = false
    open var ignoreGraphicShapeChange : Bool = false

    open override var graphic: DrawGraphic? {
        didSet {
            textContainer?.graphic = graphic
        }
    }
    
    // MARK: - Creation

    internal func initialize(from string: NSAttributedString?) {
        // Set up the NSTextStorage
        textStorage = NSTextStorage() // TODO: Maybe don't need this here.
        if let string {
            textStorage.setAttributedString(string)
        }

        // Now create and add the layout manager to the storage.
        let manager = NSLayoutManager()
        textStorage.addLayoutManager(manager)

        // And set up our primary contained.
        if let graphic {
            size = graphic.frame.size
            if size.width < 15.0 {
                size.width = 15.0
            }
            if size.height < 15.0 {
                size.height = 15.0
            }
        } else {
            size = NSSize(width: 15.0, height: 15.0)
        }
        let container = DrawTextContainer(size: size)
        container.lineFragmentPadding = lineFragmentPadding
        manager.addTextContainer(container)
        container.graphic = self.graphic
        container.heightTracksTextView = true
        container.widthTracksTextView = true

        // Finally, we'll create a text view, although we only use this when we're editing. As such, we could possibly defer the creation of it.
        let textView = DrawTextView(frame: NSRect(origin: NSPoint.zero, size: size), textContainer: container)
        textView.drawsBackground = false
        textView.isEditable = false
        textView.isSelectable = false
        textView.isHorizontallyResizable = false
        textView.isVerticallyResizable = false

        // Not going to do this just yet. I will need to do this in the future, once we allow inline editing.
        //[[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(textViewFrameDidChange:) name:NSViewFrameDidChangeNotification object:textView];
    }
    
    public required init() {
        super.init()
        initialize(from: nil)
    }

    public override init(graphic: DrawGraphic?) {
        super.init(graphic: graphic)
        initialize(from: nil)
    }

    public init(graphic: DrawGraphic, text: NSAttributedString) {
        super.init(graphic: graphic)
        initialize(from: text)
    }

    open func prepareTextInLayoutManager() {
        if let layoutManager, let textContainer {
            let range = layoutManager.glyphRange(for: textContainer)
            layoutManager.ensureLayout(forGlyphRange: range)
        }
    }

    @objc(yOffsetInTextFrame:)
    open func yOffset(in textFrame: NSRect) -> CGFloat {
        if let graphic {
            let frame = graphic.frame
            switch verticalAlignment {
            case .top:
                // Do nothing, we're automatically aligned to the top.
                return 0.0
            case .middle:
                return (frame.size.height - textFrame.size.height) / 2.0
            case .bottom:
                return frame.size.height - textFrame.size.height
            }
        }
        // Just in case we fallthrough, which we shouldn't, unless we change the enumeration and ignore warnings.
        return 0.0
    }

    open override func draw(_ path: AJRBezierPath, with priority: DrawAspectPriority) -> DrawGraphicCompletionBlock? {
        if !editing {
            if let layoutManager, let graphic, let textContainer {
                var point = graphic.frame.origin
                let range = layoutManager.glyphRange(for: textContainer)

                point.y += yOffset(in: layoutManager.usedRect(for: textContainer))

                layoutManager.drawBackground(forGlyphRange: range, at: point)
                layoutManager.drawGlyphs(forGlyphRange: range, at: point)
                
                if UserDefaults[.debugTextContainerFrames]! {
                    NSColor.red.set()
                    for container in layoutManager.textContainers {
                        NSRect(origin: point, size: container.containerSize).frame()
                    }
                }
            }
        }
        
        return nil
    }

    open var attributedString : NSAttributedString {
        get {
            return textStorage
        }
        set {
            let current = attributedString
            graphic?.document?.registerUndo(withTarget: self, handler: { object in
                self.attributedString = current
            })
            textStorage.setAttributedString(newValue)
            graphic?.setNeedsDisplay()
        }
    }

    internal func updateMaxSize() {
        if let graphic,
           let document = graphic.document,
           let textView {
            let paperSize = document.paper.size(for: document.orientation)
            let origin = graphic.frame.origin
            let textView = textView
            var maxSize = NSSize.zero

            maxSize.width = paperSize.width - document.margins.right - origin.x
            maxSize.height = paperSize.height - document.margins.bottom - origin.y
            AJRLog.debug("maxSize = \(maxSize)")
            textView.maxSize = maxSize
        }
    }

    open override func didAdd(to document: DrawDocument) {
        updateMaxSize()
        super.didAdd(to: document)
    }

    open override func willRemove(from document: DrawDocument) {
        super.willRemove(from: document)
    }

    open func updateContainerSize() {
        if let graphic,
           let textContainer {
            let graphicFrame = graphic.frame
            if let textView {
                let textFrame = textView.frame
                if textFrame != graphicFrame  {
                    textView.frame = graphicFrame
                    //[textContainer graphicDidChangeShape:aGraphic];
                }
            }
              
            if !textContainer.isSimpleRectangularTextContainer {
                textContainer.graphicDidChangeShape(graphic)
            }
            if textContainer.size != graphicFrame.size {
                var size = graphicFrame.size
                if size.width < 15.0 {
                    size.width = 15.0
                }
                if size.height < 15.0 {
                    size.height = 15.0
                }
                textContainer.containerSize = size
            }
            updateMaxSize()
        }
    }
    
    open override func graphicDidChangeShape(_ graphic: DrawGraphic) {
        if !ignoreGraphicShapeChange {
            updateContainerSize()
        }
    }
    
    open override func didAdd(to graphic: DrawGraphic) {
        updateContainerSize()
    }

    open override var aspectAcceptsEdit : Bool {
        return true
    }

    open func setupTextView(_ textView: NSTextView) {
        if let graphic,
           let container = textContainer {
            let frame = graphic.frame
            let bounds = frame

            textView.frame = frame
            textView.bounds = bounds
            textView.isEditable = true
            textView.isSelectable = true
            textView.isHorizontallyResizable = true
            textView.isVerticallyResizable = true
            container.heightTracksTextView = true
            container.widthTracksTextView = true
            updateMaxSize()
            textView.sizeToFit()
        }
    }

    open override func beginEditing(from event: DrawEvent) -> Bool {
        if let textView,
           let layoutManager,
           let graphic,
           let textContainer {
            var distance : CGFloat = 0.0
            var range = NSRange(location: 0, length: 0)

            setupTextView(textView)
            graphic.page?.addSubview(textView)
            textView.window?.makeFirstResponder(textView)
            let origin = graphic.frame.origin
            var point = event.locationOnPage
            point.x -= origin.x
            point.y -= origin.y
            range.location = layoutManager.glyphIndex(for: point, in: textContainer, fractionOfDistanceThroughGlyph: &distance)
            if distance > 0.5 {
                range.location += 1
            }
            textView.setSelectedRange(range)

            editing = true

            return true
        }
        return false
    }

    open override func endEditing() {
        if let textView, editing {
            textView.isEditable = false
            textView.isSelectable = false
            textView.isHorizontallyResizable = false
            textView.isVerticallyResizable = false
            textView.removeFromSuperview()

            editing = false

            NotificationCenter.default.post(name: .DrawObjectDidResignRuler, object: graphic?.document)
        }
    }

    open override func isPoint(_ point: NSPoint, in path: AJRBezierPath, with priority: DrawAspectPriority) -> Bool {
        if let manager = layoutManager,
           let graphic,
           let textContainer {
            // Make sure we're ready to use.
            prepareTextInLayoutManager()

            // Get the frame for our graphic, but remember to offset it by our frame.
            let graphicFrame = graphic.frame
            var rect = manager.usedRect(for: textContainer)
            rect.origin.x += graphicFrame.origin.x
            rect.origin.y += graphicFrame.origin.y + yOffset(in: rect)

            return NSMouseInRect(point, rect, false)
        }
        return false
    }

    // MARK: - DrawAspect

    open override class func defaultAspect(for graphic: DrawGraphic) -> DrawAspect? {
        return DrawText(graphic: graphic)
    }

    // MARK: - NSCopying

    internal func createTextStorage(with string: NSAttributedString) -> NSTextStorage {
        return finishInitializingTextStorage(NSTextStorage(), with: string)
    }

    internal func finishInitializingTextStorage(_ storage: NSTextStorage, with string: NSAttributedString) -> NSTextStorage {
        storage.setAttributedString(string)
        let manager = NSLayoutManager()
        storage.addLayoutManager(manager)
        let container = DrawTextContainer(containerSize: NSSize(width: 100, height: 100))
        container.lineFragmentPadding = lineFragmentPadding
        container.heightTracksTextView = true
        container.widthTracksTextView = true
        manager.addTextContainer(container)

        return storage
    }

    open override func copy(with zone: NSZone? = nil) -> Any {
        let new = super.copy(with: zone) as! DrawText

        // NOTE: Apparently copying an NSTextStorage produces an attributed string, not a text storage.
        new.textStorage = createTextStorage(with: textStorage.copy() as! NSAttributedString)
        new.editing = false
        new.lineFragmentPadding = lineFragmentPadding

        return new
    }

    // MARK: - AJRXMLArchiving

    class open override var ajr_nameForXMLArchiving: String {
        return "text"
    }

    open override func encode(with coder: AJRXMLCoder) {
        super.encode(with: coder)

        coder.encode(textStorage, forKey: "text")
        if editing {
            coder.encode(editing, forKey: "editing")
        }
        coder.encode(lineFragmentPadding, forKey:"lineFragmentPadding")
        if verticalAlignment != .top {
            coder.encode(verticalAlignment, forKey: "verticalAlignment")
        }
    }

    open override func decode(with coder: AJRXMLCoder) {
        super.decode(with: coder)

        coder.decodeObject(forKey: "text") { object in
            // Make sure the input is actually an attributed string, since it might not be if someone fiddled with the document.
            if let string = object as? NSAttributedString {
                self.textStorage = self.createTextStorage(with: string)
            } else {
                self.textStorage = self.createTextStorage(with: NSAttributedString())
            }
        }
        coder.decodeBool(forKey: "editing") { value in
            self.editing = value
        }
        coder.decodeCGFloat(forKey: "lineFragmentPadding") { value in
            self.lineFragmentPadding = value
        }
        coder.decodeEnumeration(forKey: "verticalAlignment") { (value : VerticalTextAlignment?) in
            self.verticalAlignment = value ?? .top
        }
    }

    open override func finalizeXMLDecoding() throws -> Any {
        let save = lineFragmentPadding

        initialize(from: textStorage)
        lineFragmentPadding = 0.0
        lineFragmentPadding = save
        
        return self
    }

}

// MARK: - Enumeration Functions

@_cdecl("DrawVerticalTextAlignmentFromString")
public func DrawVerticalTextAlignmentFromString(_ string: String) -> DrawText.VerticalTextAlignment {
    return DrawText.VerticalTextAlignment(string: string) ?? .top
}

@_cdecl("DrawStringFromVerticalTextAlignment")
func DrawStringFromVerticalTextAlignment(_ alignment: DrawText.VerticalTextAlignment) -> String {
    return alignment.description
}
